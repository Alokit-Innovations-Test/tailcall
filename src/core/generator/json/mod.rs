mod field_base_url_generator;
mod http_directive_generator;
mod mutation_generator;
mod query_generator;
mod schema_generator;
mod types_generator;
mod url_utils;

pub use field_base_url_generator::FieldBaseUrlGenerator;
pub use mutation_generator::OperationTypeGenerator;
pub use schema_generator::SchemaGenerator;
pub use types_generator::GraphQLTypesGenerator;
