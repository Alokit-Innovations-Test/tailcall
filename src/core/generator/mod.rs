mod from_json;
pub mod config;
mod from_proto;
mod generator;
mod graphql_type;
mod json;
mod proto;
mod from_generalized_config;
pub mod source;

pub use from_json::{from_json, ConfigGenerationRequest};
pub use json::NameGenerator;
pub use generator::Generator;
pub use from_generalized_config::FromGeneralizedConfig;
