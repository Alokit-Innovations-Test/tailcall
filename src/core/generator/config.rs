use std::path::Path;

use schemars::JsonSchema;
use serde::{Deserialize, Serialize};

use crate::core::config::{self, reader::ConfigReader};

#[derive(Serialize, Deserialize, Debug, JsonSchema)]
#[serde(rename_all = "camelCase")]
pub enum InputSource {
    /// src: maintains the same src written in config.
    /// resolved_src: holds the correctly resolved src with respect to config.
    Config {
        src: String,
        resolved_src: Option<String>,
    },
    Import {
        src: String,
        resolved_src: Option<String>,
    },
}

#[derive(Serialize, Deserialize, Debug, JsonSchema)]
#[serde(rename_all = "camelCase")]
pub struct Input {
    #[serde(flatten)]
    pub source: InputSource,
}

#[derive(Serialize, Deserialize, Default, Debug, JsonSchema)]
#[serde(rename_all = "camelCase")]
pub struct Output {
    /// Controls the output format (graphql, json, yaml)
    pub format: config::Source,
    /// Specifies the output file name
    pub file: String,
}

#[derive(Serialize, Deserialize, Debug, JsonSchema)]
#[serde(rename_all = "camelCase")]
pub struct Schema {
    #[serde(default = "defaults::schema::query")]
    pub query: String,
    #[serde(default = "defaults::schema::mutation")]
    pub mutation: String,
}

impl Default for Schema {
    fn default() -> Self {
        Self {
            query: defaults::schema::query(),
            mutation: defaults::schema::mutation(),
        }
    }
}

#[derive(Serialize, Deserialize, Default, Debug, JsonSchema)]
#[serde(rename_all = "camelCase")]
pub struct GenerateOptions {
    #[serde(default)]
    pub schema: Schema,
}

#[derive(Serialize, Deserialize, Default, Debug, JsonSchema)]
#[serde(rename_all = "camelCase")]
pub struct Transform {
    // TODO: change types
    pub ambiguous_name_resolver: Option<serde_json::Value>,
    pub merge_types: Option<serde_json::Value>,
    pub js: Option<serde_json::Value>,
}

#[derive(Serialize, Deserialize, Default, Debug, JsonSchema)]
#[serde(rename_all = "camelCase")]
pub struct GeneratorConfig {
    pub input: Vec<Input>,
    #[serde(default)]
    pub output: Output,
    #[serde(default)]
    pub generate: GenerateOptions,
    #[serde(default)]
    pub transform: Transform,
}

impl GeneratorConfig {
    /// Resolves all the paths present inside the GeneratorConfig.
    pub fn resolve_paths(mut self, file_path: &str) -> Self {
        let config_path = Path::new(file_path).parent().unwrap_or(Path::new(""));

        for input in self.input.iter_mut() {
            match &mut input.source {
                InputSource::Config { src, resolved_src }
                | InputSource::Import { src, resolved_src } => {
                    *resolved_src = Some(ConfigReader::resolve_path(src, Some(config_path)));
                }
            }
        }

        self.output.file = ConfigReader::resolve_path(&self.output.file, Some(config_path));

        self
    }
}

mod defaults {
    pub mod schema {
        pub fn query() -> String {
            "Query".to_string()
        }

        pub fn mutation() -> String {
            "Mutation".to_string()
        }
    }
}

#[cfg(test)]
mod tests {
    use insta::assert_debug_snapshot;

    use super::GeneratorConfig;

    #[test]
    fn test_from_json() {
        let content = std::fs::read_to_string(tailcall_fixtures::generator::SIMPLE_JSON).unwrap();
        let config: GeneratorConfig = serde_json::from_str(&content).unwrap();

        assert_debug_snapshot!(&config);
    }

    #[test]
    fn test_resolve_paths() {
        let file_path = "../../../tailcall-fixtures/fixtures/generator/simple-json.json";
        let content = std::fs::read_to_string(tailcall_fixtures::generator::SIMPLE_JSON).unwrap();
        let config: GeneratorConfig = serde_json::from_str(&content).unwrap();
        let config = config.resolve_paths(file_path);
        assert_debug_snapshot!(&config);
    }
}
