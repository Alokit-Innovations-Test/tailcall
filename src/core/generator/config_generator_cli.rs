use std::fs;

use inquire::Confirm;

use super::config::GeneratorConfig;
use super::source::ConfigSource;
use super::Generator;
use crate::core::config::{self, ConfigModule};
use crate::core::runtime::TargetRuntime;

/// Checks if file or folder already exists or not.
fn is_exists(path: &str) -> bool {
    fs::metadata(path).is_ok()
}

/// FIXME: move this to CLI
pub struct ConfigConsoleGenerator {
    config_path: String,
    runtime: TargetRuntime,
    generator: Generator,
}

impl ConfigConsoleGenerator {
    pub fn new(config_path: &str, runtime: TargetRuntime) -> Self {
        Self {
            generator: Generator::new(runtime.clone()),
            config_path: config_path.to_string(),
            runtime,
        }
    }

    /// Writes the configuration to the output file if allowed.
    async fn write(self, graphql_config: ConfigModule, output_path: &str) -> anyhow::Result<()> {
        let output_source = config::Source::detect(output_path)?;
        let config = match output_source {
            config::Source::Json => graphql_config.to_json(true)?,
            config::Source::Yml => graphql_config.to_yaml()?,
            config::Source::GraphQL => graphql_config.to_sdl(),
        };

        if self.should_overwrite(output_path)? {
            self.runtime
                .file
                .write(output_path, config.as_bytes())
                .await?;

            tracing::info!("Config successfully generated at {output_path}");
        }

        Ok(())
    }

    /// Checks if the output file already exists and prompts for overwrite
    /// confirmation.
    fn should_overwrite(&self, output_path: &str) -> anyhow::Result<bool> {
        if is_exists(output_path) {
            let should_overwrite = Confirm::new(
                format!(
                    "The output file '{}' already exists. Do you want to overwrite it?",
                    output_path
                )
                .as_str(),
            )
            .with_default(false)
            .prompt()?;
            if !should_overwrite {
                return Ok(false);
            }
        }
        Ok(true)
    }

    async fn read(&self) -> anyhow::Result<GeneratorConfig> {
        let config_path = &self.config_path;
        let source = ConfigSource::detect(config_path)?;
        let config_content = self.runtime.file.read(config_path).await?;

        let config: GeneratorConfig = match source {
            ConfigSource::Json => serde_json::from_str(&config_content)?,
            ConfigSource::Yml => serde_yaml::from_str(&config_content)?,
        };

        // While reading resolve the internal paths of generalized config.
        Ok(config.resolve_paths(config_path))
    }

    /// Reads the configuration from the specified path.
    pub async fn generate(self) -> anyhow::Result<()> {
        let config = self.read().await?;
        let path = config.output.file.to_owned();
        let config = self.generator.run(config.clone()).await?;

        self.write(config, &path).await?;
        Ok(())
    }
}
