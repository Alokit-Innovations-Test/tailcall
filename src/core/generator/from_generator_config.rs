use std::fs;

use inquire::Confirm;

use super::config::GeneratorConfig;
use super::source::ConfigSource;
use super::Generator;
use crate::core::config::{self, ConfigModule};
use crate::core::runtime::TargetRuntime;

/// Checks if file or folder already exists or not.
fn is_exists(path: &str) -> bool {
    fs::metadata(path).is_ok()
}

pub struct Writer {
    pub config: ConfigModule,
    runtime: TargetRuntime,
    output_path: String,
}

impl Writer {
    pub fn new(output_path: String, config: ConfigModule, runtime: TargetRuntime) -> Self {
        Self { config, runtime, output_path }
    }

    /// Checks if the output file already exists and prompts for overwrite
    /// confirmation.
    pub fn should_overwrite(&self, output_path: &str) -> anyhow::Result<bool> {
        if is_exists(output_path) {
            let should_overwrite = Confirm::new(
                format!(
                    "The output file '{}' already exists. Do you want to overwrite it?",
                    output_path
                )
                .as_str(),
            )
            .with_default(false)
            .prompt()?;
            if !should_overwrite {
                return Ok(false);
            }
        }
        Ok(true)
    }

    /// Writes the configuration to the output file if allowed.
    pub async fn write(self) -> anyhow::Result<()> {
        let output_path = &self.output_path;

        let output_source = config::Source::detect(output_path)?;
        let config = match output_source {
            config::Source::Json => self.config.to_json(true)?,
            config::Source::Yml => self.config.to_yaml()?,
            config::Source::GraphQL => self.config.to_sdl(),
        };

        if self.should_overwrite(output_path)? {
            self.runtime
                .file
                .write(output_path, config.as_bytes())
                .await?;

            tracing::info!("Config successfully generated at {output_path}");
        }

        Ok(())
    }
}

pub struct ConfigGenerator {
    config: GeneratorConfig,
    runtime: TargetRuntime,
    generator: Generator,
}

impl ConfigGenerator {
    pub fn new(config: GeneratorConfig, runtime: TargetRuntime) -> Self {
        Self { generator: Generator::new(runtime.clone()), config, runtime }
    }
    /// generates the actual configuration from generator config.
    pub async fn generate(self) -> anyhow::Result<Writer> {
        let output_path = self.config.output.file.to_owned();
        let output_config = self.generator.run(self.config).await?;
        Ok(Writer::new(output_path, output_config, self.runtime))
    }
}

pub struct Reader {
    runtime: TargetRuntime,
    config_path: String,
}

impl Reader {
    pub fn new(runtime: TargetRuntime, config_path: &str) -> Self {
        Self { runtime, config_path: config_path.to_string() }
    }

    /// Reads the configuration from the specified path.
    pub async fn generate(self) -> anyhow::Result<()> {
        let config_path = &self.config_path;
        let source = ConfigSource::detect(config_path)?;
        let config_content = self.runtime.file.read(config_path).await?;

        let config: GeneratorConfig = match source {
            ConfigSource::Json => serde_json::from_str(&config_content)?,
            ConfigSource::Yml => serde_yaml::from_str(&config_content)?,
        };

        // while reading resolve the internal paths of generalized config.
        let config = config.resolve_paths(config_path);

        ConfigGenerator::new(config, self.runtime)
            .generate()
            .await?
            .write()
            .await?;
        Ok(())
    }
}
