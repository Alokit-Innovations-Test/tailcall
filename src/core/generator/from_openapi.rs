use oas3::{OpenApiV3Spec, Spec};

use crate::core::config::Config;
use crate::core::generator::json;
use crate::core::transform::Transform;
use crate::core::valid::{Valid, Validator};

#[derive(Default)]
pub struct FromOpenAPIGenerator {
    query: String,
    #[allow(unused)]
    spec: Spec,
}

impl FromOpenAPIGenerator {
    pub fn new(query: String, spec: OpenApiV3Spec) -> Self {
        Self { query, spec }
    }
}

impl Transform for FromOpenAPIGenerator {
    type Value = Config;
    type Error = String;

    fn transform(&self, value: Self::Value) -> Valid<Self::Value, Self::Error> {
        json::SchemaGenerator::new(self.query.clone()).transform(value)
    }
}

pub fn from_openapi_spec(query: &str, spec: OpenApiV3Spec) -> Config {
    let config = Config::default();
    let final_config = FromOpenAPIGenerator::new(query.to_string(), spec)
        .transform(config)
        .to_result();
    final_config.unwrap_or_else(|e| {
        tracing::warn!("Failed to generate config from OpenAPI spec: {}", e);
        Config::default()
    })
}

#[cfg(test)]
mod tests {
    use std::path::Path;

    use super::*;

    #[test]
    fn test_openapi_apis_guru() {
        let apis_guru = config_from_openapi_spec("apis-guru.yml");
        insta::assert_snapshot!(apis_guru);
    }

    #[test]
    fn test_openapi_jsonplaceholder() {
        let jsonplaceholder = config_from_openapi_spec("jsonplaceholder.yml");
        insta::assert_snapshot!(jsonplaceholder);
    }

    #[test]
    fn test_openapi_spotify() {
        let spotify = config_from_openapi_spec("spotify.yml");
        insta::assert_snapshot!(spotify);
    }

    fn config_from_openapi_spec(filename: &str) -> String {
        let spec_path = Path::new("src")
            .join("core")
            .join("generator")
            .join("tests")
            .join("fixtures")
            .join("openapi")
            .join(filename);

        let spec = oas3::from_path(spec_path).unwrap();
        from_openapi_spec("Query", spec).to_sdl()
    }
}
