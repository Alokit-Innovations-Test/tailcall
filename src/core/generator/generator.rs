use anyhow::Result;
use futures_util::future::join_all;
use prost_reflect::prost_types::FileDescriptorSet;
use prost_reflect::DescriptorPool;

use crate::core::generator::from_proto::from_proto;
use crate::core::generator::source::ImportSource;
use crate::core::merge_right::MergeRight;
use crate::core::proto_reader::ProtoReader;
use crate::core::resource_reader::ResourceReader;
use crate::core::runtime::TargetRuntime;
use crate::core::{
    config::{self, Config, ConfigModule, Link, LinkType, Resolution},
    resource_reader::Cached,
};

use super::config::{GeneratorConfig, InputSource};

// this function resolves all the names to fully-qualified syntax in descriptors
// that is important for generation to work
// TODO: probably we can drop this in case the config_reader will use
// protox::compile instead of more low-level protox_parse::parse
fn resolve_file_descriptor_set(descriptor_set: FileDescriptorSet) -> Result<FileDescriptorSet> {
    let descriptor_set = DescriptorPool::from_file_descriptor_set(descriptor_set)?;
    let descriptor_set = FileDescriptorSet {
        file: descriptor_set
            .files()
            .map(|file| file.file_descriptor_proto().clone())
            .collect(),
    };

    Ok(descriptor_set)
}

pub struct Generator {
    runtime: TargetRuntime,
    reader: ResourceReader<Cached>,
    proto_reader: ProtoReader,
}

impl Generator {
    pub fn new(runtime: TargetRuntime) -> Self {
        let reader = ResourceReader::cached(runtime.clone());

        Self {
            runtime: runtime.clone(),
            reader: reader.clone(),
            proto_reader: ProtoReader::init(reader, runtime),
        }
    }

    pub async fn run(&self, gen_config: GeneratorConfig) -> Result<()> {
        let resolvers = gen_config.input.into_iter().map(|input| async {
            match input.source {
                InputSource::Config { src } => {
                    let source = config::Source::detect(&src)?;
                    let schema = self.reader.read_file(&src).await?;

                    Config::from_source(source, &schema.content)
                }
                InputSource::Import { src } => {
                    let source = ImportSource::detect(&src)?;

                    match source {
                        ImportSource::Proto => {
                            let metadata = self.proto_reader.read(&src).await?;
                            let descriptor_set =
                                resolve_file_descriptor_set(metadata.descriptor_set)?;
                            let mut config = from_proto(&[descriptor_set], "Query")?;

                            config.links.push(Link {
                                id: None,
                                src: metadata.path,
                                type_of: LinkType::Protobuf,
                            });

                            Ok(config)
                        }
                    }
                }
            }
        });

        let mut config = Config::default();
        for result in join_all(resolvers).await {
            config = config.merge_right(result?)
        }

        let config = ConfigModule::from(config).resolve_ambiguous_types(|v| Resolution {
            input: format!("{}Input", v),
            output: v.to_owned(),
        });

        let output_source = config::Source::detect(&gen_config.output.file)?;

        let config = match output_source {
            config::Source::Json => config.to_json(true)?,
            config::Source::Yml => config.to_yaml()?,
            config::Source::GraphQL => config.to_sdl(),
        };

        self.runtime
            .file
            .write(&gen_config.output.file, config.as_bytes())
            .await?;

        Ok(())
    }

    pub async fn read_all<T: AsRef<str>>(
        &self,
        input_source: ImportSource,
        files: &[T],
        query: &str,
    ) -> Result<ConfigModule> {
        let mut links = vec![];
        let proto_metadata = self.proto_reader.read_all(files).await?;

        let mut config = Config::default();
        for metadata in proto_metadata {
            match input_source {
                ImportSource::Proto => {
                    links.push(Link { id: None, src: metadata.path, type_of: LinkType::Protobuf });
                    let descriptor_set = resolve_file_descriptor_set(metadata.descriptor_set)?;
                    config = config.merge_right(from_proto(&[descriptor_set], query)?);
                }
            }
        }

        config.links = links;
        Ok(
            ConfigModule::from(config).resolve_ambiguous_types(|v| Resolution {
                input: format!("{}Input", v),
                output: v.to_owned(),
            }),
        )
    }
}

#[cfg(test)]
mod test {
    use std::path::PathBuf;

    use tailcall_fixtures::protobuf;

    use super::*;

    fn start_mock_server() -> httpmock::MockServer {
        httpmock::MockServer::start()
    }

    #[tokio::test]
    async fn test_read_all() {
        let server = start_mock_server();
        let runtime = crate::core::runtime::test::init(None);
        let test_dir = PathBuf::from(tailcall_fixtures::protobuf::SELF);

        let news_content = runtime.file.read(protobuf::NEWS).await.unwrap();
        let greetings_a = runtime.file.read(protobuf::GREETINGS_A).await.unwrap();

        server.mock(|when, then| {
            when.method(httpmock::Method::GET).path("/news.proto");
            then.status(200)
                .header("Content-Type", "application/vnd.google.protobuf")
                .body(&news_content);
        });

        server.mock(|when, then| {
            when.method(httpmock::Method::GET)
                .path("/greetings_a.proto");
            then.status(200)
                .header("Content-Type", "application/protobuf")
                .body(&greetings_a);
        });

        let generator = Generator::new(runtime);
        let news = format!("http://localhost:{}/news.proto", server.port());
        let greetings_a = format!("http://localhost:{}/greetings_a.proto", server.port());
        let greetings_b = test_dir
            .join("greetings_b.proto")
            .to_str()
            .unwrap()
            .to_string();

        let config = generator
            .read_all(ImportSource::Proto, &[news, greetings_a, greetings_b], "Query")
            .await
            .unwrap();

        assert_eq!(config.links.len(), 3);
        assert_eq!(config.types.get("Query").unwrap().fields.len(), 8);
    }
}
