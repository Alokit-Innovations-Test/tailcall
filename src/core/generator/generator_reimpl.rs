use prost_reflect::prost_types::FileDescriptorSet;
use prost_reflect::DescriptorPool;
use serde_json::Value;
use url::Url;

use super::from_proto::from_proto;
use super::{from_json, ConfigGenerationRequest, NameGenerator};
use crate::core::config::{Config, ConfigModule, Link, LinkType, Source};
use crate::core::merge_right::MergeRight;
use crate::core::proto_reader::ProtoMetadata;

// this function resolves all the names to fully-qualified syntax in descriptors
// that is important for generation to work
// TODO: probably we can drop this in case the config_reader will use
// protox::compile instead of more low-level protox_parse::parse
fn resolve_file_descriptor_set(
    descriptor_set: FileDescriptorSet,
) -> anyhow::Result<FileDescriptorSet> {
    let descriptor_set = DescriptorPool::from_file_descriptor_set(descriptor_set)?;
    let descriptor_set = FileDescriptorSet {
        file: descriptor_set
            .files()
            .map(|file| file.file_descriptor_proto().clone())
            .collect(),
    };

    Ok(descriptor_set)
}

pub enum GeneratorInput {
    Config { schema: String, source: Source },
    Proto { metadata: ProtoMetadata },
    Json { url: Url, data: Value },
}

pub struct Generator {
    field_name_gen: NameGenerator,
    type_name_gen: NameGenerator,
}

impl Generator {
    pub fn new(field_prefix: &str, type_prefix: &str) -> Self {
        Self {
            field_name_gen: NameGenerator::new(field_prefix),
            type_name_gen: NameGenerator::new(type_prefix),
        }
    }
    pub fn run(
        &self,
        query: &str,
        config_generation_req: &[GeneratorInput],
    ) -> anyhow::Result<ConfigModule> {
        let mut config = Config::default();

        for req in config_generation_req {
            match req {
                GeneratorInput::Config { schema, source } => {
                    config = config.merge_right(Config::from_source(source.to_owned(), schema)?)
                }
                GeneratorInput::Json { url, data } => {
                    let req = ConfigGenerationRequest::new(url.to_owned(), data.to_owned());
                    config = config.merge_right(from_json(
                        &[req],
                        query,
                        &self.field_name_gen,
                        &self.type_name_gen,
                    )?);
                }
                GeneratorInput::Proto { metadata } => {
                    let descriptor_set =
                        resolve_file_descriptor_set(metadata.descriptor_set.to_owned())?;
                    config = config.merge_right(from_proto(&[descriptor_set], query)?);

                    config.links.push(Link {
                        id: None,
                        src: metadata.path.to_owned(),
                        type_of: LinkType::Protobuf,
                    });
                }
            }
        }
        Ok(ConfigModule::from(config))
    }
}
