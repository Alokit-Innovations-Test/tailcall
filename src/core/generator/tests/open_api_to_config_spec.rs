use std::fs;
use std::path::Path;

use serde::{Deserialize, Serialize};
use serde_json::Value;
use tailcall::core::generator::from_openapi_spec;

#[derive(Serialize, Deserialize)]
struct OpenAPIFixture {
    url: String,
    body: Value,
}

pub fn run_open_api_to_config_spec(path: &Path) -> datatest_stable::Result<()> {
    let name = path
        .file_name()
        .ok_or(anyhow::anyhow!("Invalid path"))?
        .to_str()
        .ok_or(anyhow::anyhow!("Invalid string"))?;
    let name = name.strip_suffix(".yml").unwrap();
    let content = fs::read_to_string(path)?;
    test_spec(name, content)?;
    Ok(())
}

fn test_spec(snapshot_name: &str, content: String) -> anyhow::Result<()> {
    let spec = oas3::from_reader(content.as_bytes())?;
    let config = from_openapi_spec("Query", spec);
    insta::assert_snapshot!(snapshot_name, config.to_sdl());
    Ok(())
}
