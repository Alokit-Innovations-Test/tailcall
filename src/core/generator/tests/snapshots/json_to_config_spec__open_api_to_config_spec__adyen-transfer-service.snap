---
source: src/core/generator/tests/open_api_to_config_spec.rs
expression: config.to_sdl()
---
schema @server @upstream {
  query: Query
}

type Query {
  """
  >Versions 1 and 2 of the Transfers API are deprecated. If you are just starting your implementation, use the latest version.

Returns all the transactions related to a balance account, account holder, or balance platform.

When making this request, you must include at least one of the following:
- `balanceAccountId`
- `accountHolderId`
- `balancePlatform`.

This endpoint supports cursor-based pagination. The response returns the first page of results, and returns links to the next and previous pages when applicable. You can use the links to page through the results.


  """
  getTransactions(accountHolderId: String, balanceAccountId: String, balancePlatform: String, createdSince: String!, createdUntil: String!, cursor: String, limit: Int, paymentInstrumentId: String): TransactionSearchResponse @http(baseURL: "https://balanceplatform-api-test.adyen.com/btl/v1", path: "/transactions", query: [{key: "accountHolderId", value: "{{.args.accountHolderId}}"}, {key: "balanceAccountId", value: "{{.args.balanceAccountId}}"}, {key: "balancePlatform", value: "{{.args.balancePlatform}}"}, {key: "createdSince", value: "{{.args.createdSince}}"}, {key: "createdUntil", value: "{{.args.createdUntil}}"}, {key: "cursor", value: "{{.args.cursor}}"}, {key: "limit", value: "{{.args.limit}}"}, {key: "paymentInstrumentId", value: "{{.args.paymentInstrumentId}}"}])
  """
  >Versions 1 and 2 of the Transfers API are deprecated. If you are just starting your implementation, use the latest version.

Returns a transaction.
  """
  getTransactionsId(id: String!): Transaction @http(baseURL: "https://balanceplatform-api-test.adyen.com/btl/v1", path: "/transactions/{{.args.id}}")
  """
  >Versions 1 and 2 of the Transfers API are deprecated. If you are just starting your implementation, use the latest version.

Starts a transfer request to move funds within your balance platform, or send funds to a [transfer instrument](https://docs.adyen.com/api-explorer/#/legalentity/v1/post/transferInstruments). Adyen sends the outcome of the transfer request through webhooks.

To use this endpoint, you need an additional role for your API credential and transfers must be enabled for the source balance account. Your Adyen contact will set these up for you.
  """
  postTransfers(wwwAuthenticate: String): TransferOld @http(baseURL: "https://balanceplatform-api-test.adyen.com/btl/v1", method: "POST", path: "/transfers", query: [{key: "wwwAuthenticate", value: "{{.args.wwwAuthenticate}}"}])
}
